deps/chakrashim/src/v8debug.cc

```diff
@@ -21,71 +21,715 @@
+    uv_mutex_lock(&msgMutex);
+
+    if (!this->IsEmpty()) {
+      uv_mutex_unlock(&msgMutex);
+      return;
+    }
+
+    this->waitingForMessage = true;
+    uv_mutex_unlock(&msgMutex);
+    uv_sem_wait(&newMsgSem);
+  }
+
+  void MessageQueue::SetProcessMessageFunctions(
+    JsValueRef chakraDebugObject,
+    JsValueRef processDebuggerMessageFn,
+    JsValueRef processDebugEventFn,
+    JsValueRef processShouldContinueFn) {
+    this->chakraDebugObject = chakraDebugObject;
+    this->processDebuggerMessageFn = processDebuggerMessageFn;
+    this->processDebugEventFn = processDebugEventFn;
+    this->processShouldContinueFn = processShouldContinueFn;
+  }
+
+  void MessageQueue::ProcessDebuggerMessage() {
+    if (this->isProcessingDebuggerMsg) {
+      return;
+    }
+
+    if (!this->IsEmpty()) {
+      this->isProcessingDebuggerMsg = true;
+
+      wchar_t* msg = this->PopMessage();
+
+      JsValueRef msgArg;
+      JsErrorCode errorCode = JsPointerToString(msg, wcslen(msg), &msgArg);
+      CHAKRA_VERIFY_NOERROR(errorCode);
+
+      JsValueRef args[] = { jsrt::GetUndefined(), msgArg };
+      JsValueRef responseRef;
+      errorCode = JsCallFunction(
+        this->processDebuggerMessageFn, args, _countof(args), &responseRef);
+      CHAKRA_VERIFY_NOERROR(errorCode);
+
+      JsValueType resultType;
+      errorCode = JsGetValueType(responseRef, &resultType);
+      CHAKRA_VERIFY_NOERROR(errorCode);
+
+      if (resultType == JsString) {
+        v8::Local<v8::String> v8Str =
+          MessageImpl::JsValueRefToV8String(responseRef);
+        v8::MessageImpl* msgImpl = new v8::MessageImpl(v8Str);
+        Debug::handler(*msgImpl);
+      }
+
+      this->isProcessingDebuggerMsg = false;
+    }
+  }
+
+  void MessageQueue::ProcessDebugEvent(
+    JsDiagDebugEvent debugEvent,
+    JsValueRef eventData) {
+    JsValueRef debugEventRef;
+    JsErrorCode errorCode = JsIntToNumber(debugEvent, &debugEventRef);
+    CHAKRA_VERIFY_NOERROR(errorCode);
+
+    JsValueRef args[] = { jsrt::GetUndefined(), debugEventRef, eventData };
+    JsValueRef result;
+    errorCode = JsCallFunction(
+      this->processDebugEventFn, args, _countof(args), &result);
+    CHAKRA_VERIFY_NOERROR(errorCode);
+
+    JsValueType resultType;
+    errorCode = JsGetValueType(result, &resultType);
+    CHAKRA_VERIFY_NOERROR(errorCode);
+
+    if (resultType == JsString) {
+      v8::Local<v8::String> v8Str = MessageImpl::JsValueRefToV8String(result);
+      v8::MessageImpl* msgImpl = new v8::MessageImpl(v8Str);
+      Debug::handler(*msgImpl);
+    }
+  }
+
+  bool MessageQueue::ShouldContinue(JsDiagDebugEvent debugEvent) {
+    JsValueRef debugEventRef;
+    JsErrorCode errorCode = JsIntToNumber(debugEvent, &debugEventRef);
+    CHAKRA_VERIFY_NOERROR(errorCode);
+
+    JsValueRef args[] = { jsrt::GetUndefined(), debugEventRef };
+    JsValueRef result;
+    errorCode = JsCallFunction(
+      this->processShouldContinueFn, args, _countof(args), &result);
```

Comment:
We probably have one defined in xplat branch.

---
In reply to: [81947919](https://github.com/nodejs/node-chakracore/pull/126#discussion_r81947919) [](ancestors = 81947919)